lisa/sut_orchestrator/azure/arm_template.bicep

```diff
@@ -0,0 +1,317 @@
+  }
+}
+
+func getOsDiskSharedGallery(node object) object => {
+  id: resourceId(node.subscription_id, empty(node.resource_group_name) ? 'None' : node.resource_group_name, 'Microsoft.Compute/galleries/images/versions', node.image_gallery, node.image_definition, node.image_version)
+}
+
+func getOsDiskMarketplace(node object) object => {
+  publisher: node.marketplace.publisher
+  offer: node.marketplace.offer
+  sku: node.marketplace.sku
+  version: node.marketplace.version
+}
+
+func getSPWithDiskEncryptionSet(node object) object => {
+  securityEncryptionType: node.security_profile.encryption_type
+  diskEncryptionSet: {
+    id: node.security_profile.disk_encryption_set_id
+  }
+}
+
+func getSP(node object) object => {
+  securityEncryptionType: node.security_profile.encryption_type
+}
+
+func getOSImage(node object) object => {
+  name: '${node.name}-osDisk'
+  managedDisk: {
+    storageAccountType: node.os_disk_type
+    securityProfile: ((empty(node.security_profile) || (node.security_profile.security_type != 'ConfidentialVM')) ? null : (empty(node.security_profile.disk_encryption_set_id) ? getSP(node) : getSPWithDiskEncryptionSet(node)))
+  } 
+  caching: ((node.os_disk_type == 'Ephemeral') ? 'ReadOnly' : 'ReadWrite')
+  createOption: 'FromImage'
+  diskSizeGB: node.osdisk_size_in_gb
+}
+
+func getSecurityProfile(security_profile object) object => {
+  uefiSettings: {
+    secureBootEnabled: security_profile.secure_boot
+    vTpmEnabled: true
+  }
+  securityType: security_profile.security_type
+}
+
+func getAvailabilitySetId(availability_set_name string) object => {
+  id: resourceId('Microsoft.Compute/availabilitySets', availability_set_name)
+}
+
+resource nodes_vms 'Microsoft.Compute/virtualMachines@2022-08-01' = [for i in range(0, node_count): {
+  name: nodes[i].name
+  location: nodes[i].location
+  tags: vm_tags
+  plan: nodes[i].purchase_plan
+  properties: {
+    availabilitySet: (use_availability_set ? getAvailabilitySetId(availability_set_name_value) : null)
+    hardwareProfile: {
+      vmSize: nodes[i].vm_size
+    }
+    osProfile: (((!empty(nodes[i].vhd)) && (!empty(nodes[i].vhd.vmgs_path))) ? null : getOsProfile(nodes[i].short_name, admin_username, (!empty(admin_password)), admin_password, ((!empty(admin_key_data)) && nodes[i].is_linux), getLinuxConfiguration('/home/${admin_username}/.ssh/authorized_keys', admin_key_data, empty(admin_password))))
+    storageProfile: {
+      imageReference: (((!empty(nodes[i].vhd)) && (!empty(nodes[i].vhd.vmgs_path))) ? null : (((!empty(nodes[i].vhd)) && (!empty(nodes[i].vhd.vhd_path))) ? getOsDiskVhd(nodes[i].name) : ((!empty(nodes[i].shared_gallery)) ? getOsDiskSharedGallery(nodes[i].shared_gallery) : getOsDiskMarketplace(nodes[i]))))
+      osDisk: (((!empty(nodes[i].vhd)) && (!empty(nodes[i].vhd.vmgs_path))) ? getOSDisk('${nodes[i].name}-disk') : ((nodes[i].os_disk_type == 'Ephemeral') ? getEphemeralOSImage(nodes[i]) : getOSImage(nodes[i])))
```

Comment:
Is it possible to use functions for those logic?

Reply:
I have tried to use condition ? a : b in function, seems it doesn't work. I will try again.

Reply:
I put ```((!empty(nodes[i].vhd)) && (!empty(nodes[i].vhd.vmgs_path)))`` into func

Reply:
Not only this place, all ? : logic should be moved to functions, not in the general template. Think it like code. Make it a clear hierarchy.

Reply:
when use below code, it has error during deployment

```
func getOsProfile(node object, admin_username string, admin_password string, admin_key_data string) object|null => isCvm(node) ? null 
: {
  computername: node.short_name
  adminUsername: admin_username
  adminPassword: (empty(admin_password) ? null : admin_password)
  linuxConfiguration: (((!empty(admin_key_data)) && node.is_linux) ? getLinuxConfiguration('/home/${admin_username}/.ssh/authorized_keys', admin_key_data, empty(admin_password)) : null)
}
```

Reply:
Try if it works with two level of functions. First level is to return the object, called `generateOsProfile`. The second level calls the first one, called `getOsProfile` like below,

```bicep
func getOsProfile(node object, admin_username string, admin_password string, admin_key_data string) any =>  isCvm(node) ? null: getOsProfile(node, admin_username, admin_password, admin_key_data)
```

This table shows any type is useful for dynamic types.
https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/data-types#data-type-assignability

Reply:
updated

Reply:
during the testing, I found when use null or {}, both don't work, I have raise an issue in bicep github, and wait for the response.
`func getImageReference(node object) object => isCvm(node) 
? null
: generateImageReference(node)`

when use `null`, the error is `'The provided value for the output of the user-defined function '__bicep.getImageReference' is not valid. The value '' is not part of the allowed value(s): '{}'.'",`

when use `{}`, the error is `The value of parameter imageReference.publisher is invalid`

Reply:
thanks, I tried `json('null')`, it has the same error with using `null`

`The value '' is not part of the allowed value(s): '{}'.'`

Reply:
Fixed this issue. And user-defined types (including nullable types) are only available in **ARM language version 2.0**, which uses a dict for resources instead of an array. The template is still deployable.